tests/utils/test_time_utils.py

```diff
@@ -0,0 +1,364 @@
+
+        now = datetime.datetime.now(pytz.timezone("Etc/GMT-0"))
+        test_time = time_utils.current_datetime()
+
+        assert now < test_time < (now + datetime.timedelta(milliseconds=100)), \
+            f'Unexpected time value {test_time} should be close to {now}'
+
+
+    def test_block_number_to_datetime(self):
+        """Test the block_number_to_datetime function"""
+
+        start_time = datetime.datetime.strptime('28/03/1990 05:30:42', '%d/%m/%Y %H:%M:%S')
+
+        test_cases = [
+            # test 1: block number 0 (at start time)
+            {
+                "start_time": start_time, # arbitrarily chosen date
+                "block_number": 0, # first block, should be at start_time
+                "time_between_blocks": 12, # time in seconds
+                "expected_result": start_time
+            },
+            # test 2: block number 2628000 (1 year after start)
+            {
+                "start_time": start_time, # arbitrarily chosen date
+                "block_number": 365*24*60*60/12, # block after 1 year
+                "time_between_blocks": 12, # time in seconds
+                "expected_result": start_time + datetime.timedelta(days=365)
+            },
+            # test 3: block number 69420
+            {
+                "start_time": start_time, # arbitrarily chosen date
+                "block_number": 69420,
+                "time_between_blocks": 12, # time in seconds
+                "expected_result": start_time + datetime.timedelta(seconds=69420*12)
+            },
+        ]
+
+        for test_case in test_cases:
+            block_time = time_utils.block_number_to_datetime(
+                test_case["start_time"],
+                test_case["block_number"],
+                test_case["time_between_blocks"]
+            )
+
+        assert block_time == test_case["expected_result"], f'unexpected time value {block_time}'
+
+
+    def test_yearfrac_as_datetime(self):
+        """Unit tests for the yearfrac_as_datetime function"""
+
+        # Choose an arbitrary date as start_time
+        start_time = datetime.datetime.strptime('28/03/1990 05:30:42', '%d/%m/%Y %H:%M:%S')
+
+        test_cases = [
+            # test 1: yearfrac = 6 months
+            {
+                "start_time": start_time, # arbitrarily chosen date
+                "yearfrac": 0.50, # 6 months
+                "expected_result": '26/09/1990 17:30:42'
+            }
+        ]
+
+        for test_case in test_cases:
+            yearfrac_time = time_utils.yearfrac_as_datetime(
+                test_case["start_time"],
+                test_case["yearfrac"]
+            )
+
+            assert datetime.datetime.strftime(yearfrac_time, '%d/%m/%Y %H:%M:%S') ==  \
+                test_case["expected_result"], f'unexpected time value {yearfrac_time}'
+
+
+    def test_get_yearfrac_remaining(self):
+        """Unit tests for the get_yearfrac_remaining function"""
+
+        test_cases = [
```

Comment:
if mint_time > market_time is there an error?

Reply:
Great question! There is not. Should there be? As it currently is, the function won't error out so testing for that situation will give a successful (albeit nonsensical) result

Reply:
We should probably have an error -- this should be an impossible situation. mint_time is assigned to market_time when minting happens, and market_time is a monotonically increasing variable. So mint_time should always be <= market_time.